mimium-lang/src/compiler/bytecodegen.rs

```diff
@@ -458,31 +458,47 @@ impl ByteCodeGenerator {
                         let d = self.get_destination(dst.clone(), rsize);
                         bytecodes_dst.push(VmInstruction::CallCls(fadd, argsize, rsize));
                         match rsize {
-                            0 => None,
-                            1 => Some(VmInstruction::Move(d, s)),
-                            n => Some(VmInstruction::MoveRange(d, s, n)),
+                            0 => vec![],
+                            1 => vec![VmInstruction::Move(d, s)],
+                            n => vec![VmInstruction::MoveRange(d, s, n)],
                         }
                     }
                     mir::Value::Function(_idx) => {
                         unreachable!();
                     }
                     mir::Value::ExtFunction(label, ty) => {
                         let (dst, argsize, nret) =
-                        self.prepare_extcls(funcproto, bytecodes_dst, dst, args, *label, *ty);
-                    Some(VmInstruction::CallExtCls(dst, argsize, nret))
+                            self.prepare_extcls(funcproto, bytecodes_dst, dst, args, *label, *ty);
+                        vec![VmInstruction::CallExtCls(dst, argsize, nret)]
                     }
                     _ => unreachable!(),
                 }
             }
             mir::Instruction::Closure(idxcell) => {
                 let idx = self.find(idxcell);
                 let dst = self.get_destination(dst, 1);
-                Some(VmInstruction::Closure(dst, idx))
+                vec![VmInstruction::Closure(dst, idx)]
             }
-            mir::Instruction::CloseUpValue(src) => {
-                let src = self.vregister.find_keep(src).unwrap();
-                self.vregister.get_top().0.insert(dst, MemoryRegion(src, 1));
-                Some(VmInstruction::Close(src))
+            mir::Instruction::CloseUpValues(src, ty) => {
+                // src might contain multiple upvalues (e.g. tuple)
+                let flattened = ty.flatten();
+                let base_addr = self.vregister.find_keep(src).unwrap();
+
+                let mut offset = 0;
+                let mut instructions = Vec::with_capacity(flattened.len());
+                for elem_t in flattened {
+                    let tsize = Self::word_size_for_type(elem_t);
+                    if elem_t.to_type().is_function() {
+                        self.vregister
+                            .get_top()
+                            .0
+                            .insert(dst.clone(), MemoryRegion(base_addr + offset, tsize));
```

Comment:
It seems there are multiple problems.

One apparent problem (my mistake) is that,  `dst` is overwritten here when there are multiple closures.

However, even if the tuple contains one closure, this fails. So, this should be a different problem. I'm yet to figure out what's happening here.

```
fn makecounter(){
    let x = 0.0
    let countup = | |{
        let res = x
        x = x+1.0
        res
    }
    (0.0,0.0,countup)
}
let (_,_,c2) = makecounter()
fn dsp(){
    (c2(),0.0)
}
```
```
thread 'closure_counter_tuple' panicked at mimium-lang/src/runtime/vm.rs:421:14:
range end index 10 out of range for slice of length 9
```

Reply:
That code emit bytecode like this
```
makecounter
nparams:0 nret: 3
upindexes: []  state_size: 0  
constants:  [0, 2]
instructions:
  movc       1 0
  mov        0 1
  movc       2 1
  closure    2 2
  mov        1 2
  movc       5 0
  mov        2 5
  movc       5 0
  mov        3 5
  mov        5 1
  mov        4 5
  close      4
  ret        4 3
```

It closes and returns `4` but it should close `4` and return `2 3` right?

Reply:
Oh, thanks for catching! I think it should be. I'll check how that bytecode is generated.